### PR TITLE
🐛Source Gitlab: wrap HTTPError 400 in AirbyteTracedException

### PR DESCRIPTION
**Pull request by askarpets**
## What
Avoid raising `HTTPError` with code 400 on getting `refresh_token` operation
https://github.com/airbytehq/oncall/issues/2693

## How
Raise `AirbyteTracedException` in case `refresh_token` response returned with status code of 400 and error type of invalid_grant

## Recommended reading order
1. `source_gitlab/source.py`
2. `unit_tests/test_source.py`

## 🚨 User Impact 🚨
No breaking changes

## Pre-merge Actions
<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`920b9e5e7a`](https://github.com/airbytehq/airbyte/pull/31179/commits/920b9e5e7afa0ef40ef3d77f77c06e5e9ef90c07)) - ❌

⏲️  Total pipeline duration: 02mn49s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/2693-wrap-http-error/1696849070/920b9e5e7afa0ef40ef3d77f77c06e5e9ef90c07/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:920b9e5e7afa0ef40ef3d77f77c06e5e9ef90c07)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`4f1a20d47b`](https://github.com/airbytehq/airbyte/pull/31179/commits/4f1a20d47bebca560a738760ebab2bf0b18ce151)) - ❌

⏲️  Total pipeline duration: 09mn34s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/2693-wrap-http-error/1696849930/4f1a20d47bebca560a738760ebab2bf0b18ce151/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:4f1a20d47bebca560a738760ebab2bf0b18ce151)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-gitlab test report (commit [`4f1a20d47b`](https://github.com/airbytehq/airbyte/pull/31179/commits/4f1a20d47bebca560a738760ebab2bf0b18ce151)) - ❌

⏲️  Total pipeline duration: 04mn24s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-gitlab docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ❌        |
| Code format checks                                            | ✅        |
| Validate metadata for source-gitlab                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/2693-wrap-http-error/1696851467/4f1a20d47bebca560a738760ebab2bf0b18ce151/source-gitlab/1.6.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:4f1a20d47bebca560a738760ebab2bf0b18ce151)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-gitlab test
```

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py [hunk 1]
```diff
@@ -69,7 +71,17 @@ def get_new_token_expiry_date(access_token_expires_in: int, access_token_created
         return pendulum.from_timestamp(access_token_created_at + access_token_expires_in)
 
     def refresh_access_token(self) -> Tuple[str, int, int, str]:
-        response_json = self._get_refresh_access_token_response()
+        try:
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:74]: can this be implemented in the Airbyte CDK?

**Comment by bazarnov** [comment]
The CAT is failing due to my work on https://github.com/airbytehq/oncall/issues/2928, the GSM creds for `config_oauth` should be rotated for this source.  @askarpets

File: docs/integrations/sources/gitlab.md [hunk 2]
```diff
@@ -110,7 +110,8 @@ Gitlab has the [rate limits](https://docs.gitlab.com/ee/user/gitlab_com/index.ht
 ## Changelog
 
 | Version | Date       | Pull Request                                             | Subject                                                                                    |
-| :------ | :--------- | :------------------------------------------------------- | :----------------------------------------------------------------------------------------- |
+|:--------|:-----------|:---------------------------------------------------------|:-------------------------------------------------------------------------------------------|
+| 1.6.1   | 2023-10-09 | [31179](https://github.com/airbytehq/airbyte/pull/31179) | Fix HTTP Error when token is expired                                                       |
```
Comment (bazarnov) [inline docs/integrations/sources/gitlab.md:114]: nit: Added user-friendly error message, when Refresh Token has expired.

File: airbyte-integrations/connectors/source-gitlab/unit_tests/test_source.py [hunk 3]
```diff
@@ -64,3 +64,11 @@ def test_connection_fail_due_to_config_error(mocker, api_url, deployment_env, ex
     }
     status, msg = source.check_connection(logging.getLogger(), config)
     assert (status, msg) == (False, expected_message)
+
+
+def test_connection_fail_due_to_invalid_oauth_refresh_token(oauth_config, requests_mock):
+    content = {"error": "invalid_grant", "error_description": "error_description"}
+    requests_mock.post(f"https://{oauth_config['api_url']}/oauth/token", status_code=400, json=content)
+    source = SourceGitlab()
+    status, msg = source.check_connection(logging.getLogger(), oauth_config)
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-gitlab/unit_tests/test_source.py:73]: logger.getLogger("airbyte")

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py [hunk 4]
```diff
@@ -69,7 +71,17 @@ def get_new_token_expiry_date(access_token_expires_in: int, access_token_created
...
+            response_json = self._get_refresh_access_token_response()
+        except HTTPError as e:
+            content = e.response.json()
+            if e.response.status_code == 400 and content.get("error") == "invalid_grant":
+                raise AirbyteTracedException(
+                    internal_message=content.get("error_description"),
+                    message="Refresh token is invalid or expired. Please re-authenticate to restore access to Gitlab.",
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:81]: We should also guide the Customer where to do so, like "... Please re-authenticate from Sources > your source > settings"

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:74]: I agree, that the `400` should be handled on the CDK level. However, some Data Providers could get the same error using a different `status_code`, like `500` or `401` in some cases. + I don't think we should include the error message in the CDK implementation.

**Review by bazarnov — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py (hunk 1 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py:74]: We can override status codes and messages on the connector level, what do you think?

**Pull request closed without merging**